### PR TITLE
add platform env vars for devices

### PR DESCRIPTION
**Pull request by Steve-Mcl**
part of #841

Raised (early) as draft until direction is ratified and tests are added

### CODE REVIEW

**Comment by Steve-Mcl** [comment]
See https://github.com/flowforge/flowforge/issues/841 for implementation details

File: frontend/src/pages/admin/Template/sections/Environment.vue [hunk 1]
```diff
@@ -88,8 +91,8 @@ import FormRow from '@/components/FormRow'
 import FormHeading from '@/components/FormHeading'
 import LockSetting from '../components/LockSetting'
 import ChangeIndicator from '../components/ChangeIndicator'
-import { TrashIcon, PlusSmIcon, LockClosedIcon } from '@heroicons/vue/outline'
-
+import { TrashIcon, PlusSmIcon, LockClosedIcon, CogIcon } from '@heroicons/vue/outline'
+const RESERVED_ENV = ['FF_PROJECT_ID', 'FF_PROJECT_NAME', 'FF_DEVICE_ID', 'FF_DEVICE_NAME', 'FF_DEVICE_TYPE', 'FF_SNAPSHOT_ID', 'FF_SNAPSHOT_NAME']
```
Comment (knolleary) [inline frontend/src/pages/admin/Template/sections/Environment.vue:95]: How about we consider `FF_*` to be reserved rather than try to maintain a list in multiple places?

**Review by knolleary — COMMENTED** [review]
A few comments for discussion.

File: frontend/src/pages/admin/Template/sections/Environment.vue [hunk 2]
```diff
@@ -88,8 +91,8 @@ import FormRow from '@/components/FormRow'
 import FormHeading from '@/components/FormHeading'
 import LockSetting from '../components/LockSetting'
 import ChangeIndicator from '../components/ChangeIndicator'
-import { TrashIcon, PlusSmIcon, LockClosedIcon } from '@heroicons/vue/outline'
-
+import { TrashIcon, PlusSmIcon, LockClosedIcon, CogIcon } from '@heroicons/vue/outline'
```
[the remark this answers is not in the document]
Comment (Steve-Mcl) [reply frontend/src/pages/admin/Template/sections/Environment.vue:94]: I used cog to differentiate platform defined vars.

File: forge/routes/api/device.js [hunk 3]
```diff
@@ -259,10 +259,12 @@ module.exports = async function (app) {
             let changed = false
             if (request.body.name !== undefined && request.body.name !== request.device.name) {
                 request.device.name = request.body.name
+                sendDeviceUpdate = true
                 changed = true
             }
             if (request.body.type !== undefined && request.body.type !== request.device.type) {
                 request.device.type = request.body.type
+                sendDeviceUpdate = true
```
[the remark this answers is not in the document]
Comment (Steve-Mcl) [reply forge/routes/api/device.js:267]: I included one extra EV - FF_DEVICE_TYPE since i suspect should a user change this field it is because of a HW change and thus very useful to know inside your flow what HW type is specified or expected (for things like alternate GPIO numbering etc)

File: frontend/src/pages/admin/Template/sections/Environment.vue (hunk 1 above)
Comment (Steve-Mcl) [reply frontend/src/pages/admin/Template/sections/Environment.vue:95]: ok, can do. 

will push change in a moment

File: frontend/src/pages/admin/Template/sections/Environment.vue (hunk 2 above)
[the remark this answers is not in the document]
Comment (Steve-Mcl) [reply frontend/src/pages/admin/Template/sections/Environment.vue:94]: Would you suggest another icon that indicates it is a platform setting or are happy just to re-use the LockClosedIcon?

File: frontend/src/pages/admin/Template/sections/Environment.vue (hunk 1 above)
Comment (Steve-Mcl) [reply frontend/src/pages/admin/Template/sections/Environment.vue:95]: oh - wait. Does that mean at the backend too (not just front end)

File: frontend/src/pages/admin/Template/sections/Environment.vue (hunk 2 above)
[the remark this answers is not in the document]
Comment (knolleary) [reply frontend/src/pages/admin/Template/sections/Environment.vue:94]: I suggested we reuse LockClosedIcon because from a users point-of-view it is an env var they cannot edit - so a lock icon would be consistent. I do not think the fact it is system-defined is something that needs indicating.

File: frontend/src/pages/admin/Template/sections/Environment.vue (hunk 1 above)
Comment (knolleary) [reply frontend/src/pages/admin/Template/sections/Environment.vue:95]: If we say `FF_*` are reserved, then the API should not allow a user to set any env vars starting `FF_*`.

File: frontend/src/pages/admin/Template/sections/Environment.vue (hunk 2 above)
[the remark this answers is not in the document]
Comment (Steve-Mcl) [reply frontend/src/pages/admin/Template/sections/Environment.vue]: resolved in [f1b77a1](https://github.com/flowforge/flowforge/pull/1039/commits/f1b77a1799f2adfd1e680404ec88144508817fa4)

File: frontend/src/pages/admin/Template/sections/Environment.vue (hunk 1 above)
Comment (Steve-Mcl) [reply frontend/src/pages/admin/Template/sections/Environment.vue]: resolved in [2dd72bc](https://github.com/flowforge/flowforge/pull/1039/commits/2dd72bceaab1691037fa8ba2d82c38d63da4d0d3)

**Comment by Steve-Mcl** [comment]
@knolleary  Should be as per discussion & ready for re-review.

Also, while executing updated tests locally, I found a duplicate unit test name clashing with the "Project API" on what I suspect should have been "Settings API" - see commit
[ba21cde](https://github.com/flowforge/flowforge/pull/1039/commits/ba21cde4178f6da27a14f9e5a9a7c767922c89b8)

File: forge/db/controllers/Project.js [hunk 4]
```diff
@@ -170,6 +169,35 @@ module.exports = {
         const newHash = crypto.createHash('sha256').update(newKey).digest()
         const oldHash = crypto.createHash('sha256').update(oldKey).digest()
         return encryptCreds(newHash, decryptCreds(oldHash, original))
+    },
+
+    /**
+     * Remove platform specific environment variables
+     * @param {[{name:string, value:string}]} envVars Environment variables array
+     */
+    removePlatformSpecificEnvVars: function (app, envVars) {
```
[the remark this answers is not in the document]
Comment (hardillb) [reply forge/db/controllers/Project.js:178]: Leave it for now

File: forge/db/migrations/20221008-01-update-device-settings-hash.js [hunk 5]
```diff
@@ -0,0 +1,12 @@
+/**
+ * Revert the starter teamType userLimit to 0
```
Comment (knolleary) [inline forge/db/migrations/20221008-01-update-device-settings-hash.js:2]:
```suggestion
 * Reset device settings hashes so they can be re-calculated as needed
```

**Comment by knolleary** [comment]
Looks good.

Only visual nit is if a device's type has not been set (which is valid):

image

I suspect this is coming from the template code which does have a habit of inserting 'No Value' in a few places which looks a bit odd. I'm not expecting you go off and address that wider issue - but I think we should avoid saying 'No Value' in the context of env vars and just leave the value blank.

**Comment by Steve-Mcl** [comment]
> Only visual nit is if a device's type has not been set (which is valid):
> 
> image
> 
> I suspect this is coming from the template code which does have a habit of inserting 'No Value' in a few places which looks a bit odd. I'm not expecting you go off and address that wider issue - but I think we should avoid saying 'No Value' in the context of env vars and just leave the value blank.

Is 'No Value' getting through to the device it's self Nick?

I didn't check this.

Possibly needs a test adjustment too.

**Comment by Steve-Mcl** [comment]
> 'No Value'

resolved in [0a8f458](https://github.com/flowforge/flowforge/pull/1039/commits/0a8f458494d7e9e9878ef61c530e440b5f5ce9a2)

image


image

**Review by knolleary — APPROVED** [review]
